File: myutils/src/main/java/pl/mareklangiewicz/myutils/MyTextUtils.java
package pl.mareklangiewicz.myutils;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.View;

import java.util.List;
import java.util.Locale;

/**
 * Created by Marek Langiewicz on 07.09.15.
 * Utilities for text manipulation
 */
public final class MyTextUtils {

    private MyTextUtils() {
        throw new AssertionError("MyTextUtils class is noninstantiable.");
    }

    /*
        private static final boolean V = BuildConfig.VERBOSE;
        private static final boolean VV = BuildConfig.VERY_VERBOSE;

        FIXME SOMEDAY: enable version with BuildConfig when Google fix issue with propagating build types to libraries.
        Now it is always 'release' in libraries.. see:
        https://code.google.com/p/android/issues/detail?id=52962
        http://stackoverflow.com/questions/20176284/buildconfig-debug-always-false-when-building-library-projects-with-gradle
        http://tools.android.com/tech-docs/new-build-system/user-guide#TOC-Library-Publication
    */
    private static final boolean V = true;
    private static final boolean VV = false;


    static public @NonNull String str(int x) {
        if(VV)
            return toVeryLongStr(x);
        if(V)
            return toLongStr(x);
        return toShortStr(x);
    }
    //TODO SOMEDAY: str for other primitive types..

    static public @NonNull <E> String str(@Nullable List<E> x) {
        if(VV)
            return toVeryLongStr(x);
        else if(V)
            return toLongStr(x);
        else
            return toShortStr(x);
    }

    static public @NonNull String str(@Nullable Object x) {
        if(VV)
            return toVeryLongStr(x);
        else if(V)
            return toLongStr(x);
        else
            return toShortStr(x);
    }

    static private @NonNull <E> String toStr(@Nullable List<E> list, int max) {
        if(list == null)
            return "null";
        StringBuilder result = new StringBuilder();
        result.append("[");
        int N = Math.min(list.size(), max);
        for(int i = 0; i < N; ++i) {
            result.append(str(list.get(i)));
            if(i < N - 1)
                result.append((","));
        }
        if(max < list.size())
            result.append(",...");
        result.append("]");
        return result.toString();
    }

    static public @NonNull <E> String toVeryLongStr(@Nullable List<E> list) {
        return toStr(list, 1000);
    }

    static public @NonNull <E> String toLongStr(@Nullable List<E> list) {
        return toStr(list, 100);
    }

    static public @NonNull <E> String toShortStr(@Nullable List<E> list) {
        return toStr(list, 10);
    }

    static public @NonNull String toShortStr(int x) {
        return Integer.toString(x);
    }
    //TODO SOMEDAY: toShortStr for other primitive types..

    static public @NonNull String toShortStr(@Nullable Object x) {
        throw new UnsupportedOperationException(); //TODO SOMEDAY
    }

    static public @NonNull String toLongStr(int x) {
        return Integer.toString(x);
    }
    //TODO SOMEDAY: toLongStr for other primitive types..

    static public @NonNull String toLongStr(@Nullable Object x) {

        if(x == null)
            return "null";

        //Shorten some very long android object descriptions:
        if(x instanceof Bundle)
            return String.format(Locale.US, "%s{size:%d}", x.getClass().getSimpleName(), ((Bundle) x).size());
        if(x instanceof View)
            return String.format(Locale.US, "%s{hash:%x}", x.getClass().getSimpleName(), x.hashCode());

        //TODO SOMEDAY: pretty print our android special cases.

        return x.toString();
    }

    static public @NonNull String toVeryLongStr(int x) {
        return Integer.toString(x);
    }
    //TODO SOMEDAY: toVeryLongStr for other primitive types..

    static public @NonNull String toVeryLongStr(@Nullable Object x) {

        if(x == null)
            return "null";

        //TODO SOMEDAY: very long and pretty print our android special cases.

        return x.toString();
    }

}

File: myutils/src/main/java/pl/mareklangiewicz/myutils/MyTextUtils.kt
package pl.mareklangiewicz.myutils

import android.os.Bundle
import android.view.View

private val V = true
private val VV = false

/**
 * str extension functions are for humans only. returned string format can change any time..
 */

val Double?.str: String get() = if(VV) toVeryLongStr() else if(V) toLongStr() else toShortStr()
val Float?.str: String get() = if(VV) toVeryLongStr() else if(V) toLongStr() else toShortStr()
val Long?.str: String get() = if(VV) toVeryLongStr() else if(V) toLongStr() else toShortStr()
val Int?.str: String get() = if(VV) toVeryLongStr() else if(V) toLongStr() else toShortStr()
val Short?.str: String get() = if(VV) toVeryLongStr() else if(V) toLongStr() else toShortStr()
val Byte?.str: String get() = if(VV) toVeryLongStr() else if(V) toLongStr() else toShortStr()
val Char?.str: String get() = if(VV) toVeryLongStr() else if(V) toLongStr() else toShortStr()
val Boolean?.str: String get() = if(VV) toVeryLongStr() else if(V) toLongStr() else toShortStr()
val <E> Array<E>?.str: String get() = if(VV) toVeryLongStr() else if(V) toLongStr() else toShortStr()
val <E> Iterable<E>?.str: String get() = if(VV) toVeryLongStr() else if(V) toLongStr() else toShortStr()
val Any?.str: String get() = if(VV) toVeryLongStr() else if(V) toLongStr() else toShortStr()

private fun <E> Array<E>?.toStr(max: Int) = this?.joinToString(prefix = "[", postfix = "]", limit = max) ?: "null"
private fun <E> Iterable<E>?.toStr(max: Int) = this?.joinToString(prefix = "[", postfix = "]", limit = max) ?: "null"

fun <E> Array<E>?.toShortStr() = toStr(10)
fun <E> Array<E>?.toLongStr() = toStr(100)
fun <E> Array<E>?.toVeryLongStr() = toStr(1000)

fun <E> Iterable<E>?.toShortStr() = toStr(10)
fun <E> Iterable<E>?.toLongStr() = toStr(100)
fun <E> Iterable<E>?.toVeryLongStr() = toStr(1000)

fun Double?.toShortStr() = toString() // exact string representation can change..
fun Double?.toLongStr() = toString() // exact string representation can change..
fun Double?.toVeryLongStr() = toString() // exact string representation can change..
fun Float?.toShortStr() = toString() // exact string representation can change..
fun Float?.toLongStr() = toString() // exact string representation can change..
fun Float?.toVeryLongStr() = toString() // exact string representation can change..
fun Long?.toShortStr() = toString() // exact string representation can change..
fun Long?.toLongStr() = toString() // exact string representation can change..
fun Long?.toVeryLongStr() = toString() // exact string representation can change..
fun Int?.toShortStr() = toString() // exact string representation can change..
fun Int?.toLongStr() = toString() // exact string representation can change..
fun Int?.toVeryLongStr() = toString() // exact string representation can change..
fun Short?.toShortStr() = toString() // exact string representation can change..
fun Short?.toLongStr() = toString() // exact string representation can change..
fun Short?.toVeryLongStr() = toString() // exact string representation can change..
fun Byte?.toShortStr() = toString() // exact string representation can change..
fun Byte?.toLongStr() = toString() // exact string representation can change..
fun Byte?.toVeryLongStr() = toString() // exact string representation can change..
fun Char?.toShortStr() = toString() // exact string representation can change..
fun Char?.toLongStr() = toString() // exact string representation can change..
fun Char?.toVeryLongStr() = toString() // exact string representation can change..
fun Boolean?.toShortStr() = this?.let { if(it) "t" else "f" } ?: "null" // exact string representation can change..
fun Boolean?.toLongStr() = this?.let { if(it) "true" else "false" } ?: "null" // exact string representation can change..
fun Boolean?.toVeryLongStr() = toLongStr() // exact string representation can change..

fun Any?.toShortStr() = toString() // TODO SOMEDAY: print some short id or something..
fun Any?.toLongStr() = toString()
fun Any?.toVeryLongStr() = toString() // TODO SOMEDAY: print some more info..

val Bundle?.str: String get() = if(VV) toVeryLongStr() else if(V) toLongStr() else toShortStr()

fun Bundle?.toShortStr() = this?.let { "Bundle{size:${size()}}" } ?: "null"
fun Bundle?.toLongStr() = toShortStr()
fun Bundle?.toVeryLongStr() = toString()

val View?.str: String get() = if(VV) toVeryLongStr() else if(V) toLongStr() else toShortStr()

fun View?.toShortStr() = this?.let { "View{hash:${hashCode()}}" } ?: "null"
fun View?.toLongStr() = toShortStr()
fun View?.toVeryLongStr() = toString()

